server/src/main/java/org/tctalent/server/service/db/impl/DuolingoCouponServiceImpl.java

```diff
@@ -0,0 +1,206 @@
+        if (seenCouponCodes.contains(couponCode)) continue; // Skip if already processed
+        seenCouponCodes.add(couponCode); // Mark code as processed
+
+        if (!couponRepository.existsByCouponCode(couponCode)) {
+          DuolingoCoupon coupon = new DuolingoCoupon();
+          coupon.setCouponCode(couponCode);
+          coupon.setAssigneeEmail(getNullableValue(line[1]));
+          coupon.setExpirationDate(parseDate(line[2], FORMATTER1, FORMATTER2));
+          coupon.setDateSent(parseDate(line[3], FORMATTER1, FORMATTER2));
+          coupon.setCouponStatus(getNullableValue(line[4]));
+          coupon.setTestStatus(getNullableValue(line[5]));
+          newCoupons.add(coupon);
+        }
+      }
+      if (!newCoupons.isEmpty()) couponRepository.saveAll(newCoupons);
+
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to import coupons from CSV file", e);
+    }
+  }
+
+  /**
+   * Assigns an available coupon to a candidate.
+   *
+   * @param candidateId the ID of the candidate
+   * @return the assigned Coupon if successful, or an empty Optional if not
+   */
+  @Override
+  @Transactional
+  public Optional<DuolingoCoupon> assignCouponToCandidate(Long candidateId) {
+    return candidateRepository.findById(candidateId).flatMap(candidate -> {
+      Optional<DuolingoCoupon> availableCoupon = couponRepository.findTop1ByCandidateIsNullAndCouponStatus("AVAILABLE");
+      availableCoupon.ifPresent(coupon -> {
+        coupon.setCandidate(candidate);
+        coupon.setDateSent(LocalDateTime.now());
+        coupon.setAssigneeEmail(candidate.getUser().getEmail());
+        coupon.setCouponStatus("SENT");
+        couponRepository.save(coupon);
+      });
+      return availableCoupon;
+    });
+  }
+
+  /**
+   * Retrieves coupons assigned to a candidate.
+   *
+   * @param candidateId the candidate's ID
+   * @return a list of CouponResponse objects representing the candidate's coupons
+   */
+  @Override
+  @Transactional(readOnly = true)
+  public List<DuolingoCouponResponse> getCouponsForCandidate(Long candidateId) {
+    List<DuolingoCoupon> coupons = couponRepository.findAllByCandidateId(candidateId);
+    List<DuolingoCouponResponse> couponResponses = new ArrayList<>();
+    for (DuolingoCoupon coupon : coupons) {
+      couponResponses.add(mapToCouponResponse(coupon));
+    }
+    return couponResponses;
+  }
+
+  /**
+   * Updates the status of a coupon identified by its code.
+   *
+   * @param couponCode the code of the coupon to update
+   * @param status the new status to set
+   */
+  @Override
+  @Transactional
+  public void updateCouponStatus(String couponCode, String status) {
+    couponRepository.findByCouponCode(couponCode).ifPresent(coupon -> {
+      coupon.setCouponStatus(status);
+      couponRepository.save(coupon);
+    });
+  }
+
+  /**
+   * Retrieves all available (unassigned) coupons.
+   *
+   * @return a list of available Coupon entities
+   */
```

Comment:
When implementing an interface - the interface should be documented (which it is) - and there is usually no need to JavaDoc the implementation. Having two sets of doc for the same method is redundant and potentially inconsistent.